internal/services/datafactory/data_factory_dataset_azure_sql_table_resource.go

```diff
@@ -0,0 +1,436 @@
+				if len(data.Annotations) > 0 {
+					annotations := make([]interface{}, len(data.Annotations))
+					for i, v := range data.Annotations {
+						annotations[i] = v
+					}
+					azureSqlTable.Annotations = &annotations
+				} else {
+					azureSqlTable.Annotations = nil
+				}
+			}
+
+			if metadata.ResourceData.HasChange("additional_properties") {
+				azureSqlTable.AdditionalProperties = data.AdditionalProperties
+			}
+
+			if metadata.ResourceData.HasChange("schema_column") {
+				azureSqlTable.Structure = data.SchemaColumn
+			}
+
+			dataset = datafactory.DatasetResource{
+				Type:       pointer.To(string(datafactory.TypeBasicDatasetTypeAzureSQLTable)),
+				Properties: azureSqlTable,
+			}
+
+			if _, err := client.CreateOrUpdate(ctx, id.ResourceGroup, id.FactoryName, id.Name, dataset, ""); err != nil {
+				return fmt.Errorf("updating %s: %+v", id, err)
+			}
+
+			metadata.SetID(id)
+			return nil
+		},
+	}
+}
+
+func (DataFactoryDatasetAzureSQLTableResource) Read() sdk.ResourceFunc {
+	return sdk.ResourceFunc{
+		Timeout: 5 * time.Minute,
+		Func: func(ctx context.Context, metadata sdk.ResourceMetaData) error {
+			d := metadata.ResourceData
+			client := metadata.Client.DataFactory.DatasetClient
+			id, err := parse.DataSetID(d.Id())
+			if err != nil {
+				return err
+			}
+
+			dataFactoryId := factories.NewFactoryID(id.SubscriptionId, id.ResourceGroup, id.FactoryName)
+			var state DataFactoryDatasetAzureSQLTableResourceSchema
+
+			resp, err := client.Get(ctx, id.ResourceGroup, id.FactoryName, id.Name, "")
+			if err != nil {
+				if utils.ResponseWasNotFound(resp.Response) {
+					return metadata.MarkAsGone(id)
+				}
+
+				return fmt.Errorf("retrieving %s: %+v", id, err)
+			}
+
+			state.Name = id.Name
+			state.DataFactoryId = dataFactoryId.ID()
+
+			azureSqlTable, ok := resp.Properties.AsAzureSQLTableDataset()
+			if !ok {
+				return fmt.Errorf("classifying Data Factory Dataset Azure SQL Table %s: Expected: %q Received: %T", *id, datafactory.TypeBasicDatasetTypeAzureSQLTable, resp.Properties)
+			}
+
+			state.AdditionalProperties = azureSqlTable.AdditionalProperties
+
+			if azureSqlTable.Description != nil {
+				state.Description = pointer.From(azureSqlTable.Description)
+			}
+
+			state.Parameters = flattenDataSetParameters(azureSqlTable.Parameters)
+			state.Annotations = flattenDataFactoryAnnotations(azureSqlTable.Annotations)
+
+			if linkedService := azureSqlTable.LinkedServiceName; linkedService != nil && linkedService.ReferenceName != nil {
+				state.LinkedServiceId = parse.NewLinkedServiceID(id.SubscriptionId, id.ResourceGroup, id.FactoryName, *linkedService.ReferenceName).ID()
+			}
+
+			if properties := azureSqlTable.AzureSQLTableDatasetTypeProperties; properties != nil {
+				val, ok := properties.TableName.(string)
+				if !ok {
+					log.Printf("[DEBUG] Skipping `table_name` since it's not a string")
+				} else {
+					state.TableName = val
+				}
+			}
```

Comment:
Two things here:

1. We validate the input and only ever send a string for `table_name`, if its likely for the API to send back something else entirely we should raise a bug on the REST API spec for this instead of handling it in the provider.

2. We always need to set a value into state

So this should become

```suggestion
			tableName := ""
			if properties := azureSqlTable.AzureSQLTableDatasetTypeProperties; properties != nil && properties.TableName != nil {
				tableName = properties.TableName.(string)
			}
			state.TableName = tableName
```

Reply:
Their doc says that this might be an `Expression` (similar to `azurerm_data_factory_dataset_binary`.`azure_blob_storage_location.dynamic_{container,path,filename}_enabled`), but in the ADF UI, I don't see them making this field available for a dynamic expression... Yeah right we might need to file an issue to azure-rest-api-spec.

![image](https://github.com/hashicorp/terraform-provider-azurerm/assets/12882489/ec64f3e9-5f3d-4082-bd52-5dff768327a6)

Reply:
Updated. I am also changing `table_name` to `schema` and `table` as per Azure document.
https://learn.microsoft.com/en-us/azure/templates/microsoft.datafactory/factories/datasets?pivots=deployment-language-bicep#azuresqltabledatasettypeproperties